### DISCUSSION TITLE
Combobox vs. Button

### DISCUSSION THREAD

**Discussion by jess-watson** [twilio-labs/paste#3245] [General] [answered]
Hey team,

I'm working on migrating some Evergreen components and interactions to Paste ([here](https://www.figma.com/file/TlUJ33t6zwJRtVNvLb3rQa/%5BDesign%5D-Project-Thunder?type=design&node-id=1098%3A36590&t=sttp0s4fdJxty501-1)). In other parts of the product we use a button that opens a menu of audiences, events, etc that a user can choose from, however I'm wondering if a combobox is a better component to use in this situation.

Here are some of my questions:

- Should we use a single select combobox for this interaction, or continue using a button?
- For the ‘create audience’ flow, can we have the ‘create’ button live within the listbox, or is it more accessible to have a separate button?
- The Segment app has a “two panel” menu where the right side of the menu shows information about what a user selected. Is there a way we can build this in Paste? Is this accessible? If we can’t recreate it, is there a better way to display this information?

Thanks for your input!

**Comment by loreina** [comment, marked as answer]
Hi @jess-watson,

> Should we use a single select combobox for this interaction, or continue using a button?

I think a Combobox works well here. If using a Button, you would have to use the [Listbox Primitive](https://paste.twilio.design/primitives/listbox-primitive) to build the popover. It would be simpler, if it's a single panel, to use Combobox. 

> For the ‘create audience’ flow, can we have the ‘create’ button live within the listbox, or is it more accessible to have a separate button?

Your engineers should be able to compose an option within Combobox to create options. We have it in our backlog to add an example to our docs site. 

> The Segment app has a “two panel” menu where the right side of the menu shows information about what a user selected. Is there a way we can build this in Paste? Is this accessible? If we can’t recreate it, is there a better way to display this information?

You can make the double panel accessible, but you would need to build everything in it, and it has to be a popover. You can't have that type of content in a combobox value right now. The [ARIA APG's Combobox pattern](https://www.w3.org/WAI/ARIA/apg/patterns/combobox/) accommodates this, but it doesn't work super well and we don't currently support it.

@darcimm came by to discuss the double panel component a little bit ago if you want to review the office hours recording or sync with her: https://github.com/twilio-labs/paste/discussions/3104


Hope that helps! Let us know what you decide to do with the double panel component as our team would be curious on how it turns out. Thanks!

**Reply by darcimm** [reply]
Hey there! Following up on this as I start to try to build out a similar pattern... is the Menu item in Paste something we could adapt to include a side-by-side panel?

**Reply by SiTaggart** [reply]
Hi @darcimm, unfortunately not. Components or UI controls or Widgets, whatever we want to call them, can only perform the action that it is designed to do. We are governed by a couple of things:

- The HTML spec
- ARIA and ARIA [Authoring practices](https://www.w3.org/WAI/ARIA/apg/patterns/).

Both of these describe all the known concepts of UI controls that are able to communicate their Identity, Operation and State to assistive technologies (AT). In essence, they have predefined anatomy.

We can't change or adapt that anatomy or intended use case because browsers and AT don't know how to communicate it and non-visual users don't know how to interpret them. We basically can't just make up or invent UI controls. We can, on the other hand, compose know controls into more complex compositions. So you might be able to build something similar using [Pills](https://paste.twilio.design/components/form-pill-group), [Popovers](https://paste.twilio.design/components/popover) and [Listboxes](https://paste.twilio.design/primitives/listbox-primitive)

<img width="758" alt="image" src="https://github.com/twilio-labs/paste/assets/368249/9092ce3f-66ce-4bb2-b0d4-f315481b7e6b">

**Reply by marektonik** [reply]
@SiTaggart hey! we are working on these comboboxes and wondering if we could remove popover's padding as on default it's 32px and it's a lot for these more complex comboboxes, especially having on mind that there is a max-width set. Could you please confirm if it's ok to remove this padding + give us some additional guidance? thanks!

below screenshot, here [link to figma](https://www.figma.com/file/oeUbF1qpn2FUsmXwiTsHmN/Segment-App%3A-Paste-Migration?type=design&node-id=2117-134190&mode=design&t=zZY0HirCKYDP9Vyt-4)
<img width="1818" alt="Zrzut ekranu 2023-09-8 o 17 27 48" src="https://github.com/twilio-labs/paste/assets/140410476/f5907997-8964-4efd-b649-3a12505226a0">

**Reply by SiTaggart** [reply]
Hi there

Having had another conversation about this component this week in office hours, probing into the usability, where even the implementing team had questions about the validity or need for some features, we have come to the conclusion that this component should be completely rethought as an experience and not reimplemented.

We won't be supporting it, or recommending it.

This should be broken down into it's sub parts and displayed on the page as form fields.


Thanks

**Reply by marektonik** [reply]
@SiTaggart hey, understand. And how comboboxes like this one? In addition to regular combobox component, there is a search. Should this experience be rethought also, or can we compose it like that?

<img width="393" alt="Zrzut ekranu 2023-09-11 o 17 53 49" src="https://github.com/twilio-labs/paste/assets/140410476/ab76a646-363b-43cd-b1de-64cb56872f58">

**Reply by SiTaggart** [reply]
That should just be a multi select Combobox https://paste.twilio.design/components/multiselect-combobox. No need to hide the autocomplete inside a popover, just let folks type into an input.
